modules/apps/transfer/types/transfer_authorization.go

```diff
@@ -145,6 +158,39 @@
 	return false
 }
 
+// areAllowedPacketDataKeys returns a boolean indicating if the memo is valid for transfer.
+func areAllowedPacketDataKeys(ctx sdk.Context, memo string, allowedPacketDataList []string) (bool, error) {
+	// if the allow list is empty, then the memo must be an empty string
+	if len(allowedPacketDataList) == 0 {
+		return len(strings.TrimSpace(memo)) == 0, nil
+	}
+
+	// if allowedPacketData have only 1 elements and it equal AllowAllPacketDataKeys
+	// then accept all the packet
+	if len(allowedPacketDataList) == 1 && allowedPacketDataList[0] == AllowAllPacketDataKeys {
+		return true, nil
+	}
+
+	// unmarshal memo
+	jsonObject := make(map[string]interface{})
+	err := json.Unmarshal([]byte(memo), &jsonObject)
+	if err != nil {
+		return false, err
+	}
+
+	gasCostPerIteration := ctx.KVGasConfig().IterNextCostFlat
+
+	for key := range jsonObject {
+		ctx.GasMeter().ConsumeGas(gasCostPerIteration, "transfer authorization")
+
+		if !slices.Contains(allowedPacketDataList, key) {
+			return false, fmt.Errorf("not allowed packet data key: %s", key)
+		}
+	}
```

Comment:
I think we no need to worry about this. The `jsonObject` only use here. It isn't change anything when we iteration over it

Reply:
The code will need to be readjusted. Map iteration in go is non-determinstic so if you have a key which is invalid, each node will take a different number of iterations to hit the invalid key. Since the gas is incremented for each iteration, each node could potentially have a different amount of gas used. Since the gas used is used to construct the block hash, nodes will diverge and create an app hash mismatch resulting in a chain halt

Someone from our team will take a look and come up with an alternative and safe approach